source/agora/consensus/ValidatorSet.d

```diff
@@ -0,0 +1,722 @@
+    private Database db;
+
+    /// The cycle length for a validator
+    public static immutable uint ValidatorCycle = 1008; // freezing period / 2
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            db_path = path to the database file, or in-memory storage if
+                        :memory: was passed
+
+    ***************************************************************************/
+
+    public this (string db_path)
+    {
+        this.db = Database(db_path);
+
+        // create the table for validator set if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS validator_set " ~
+            "(key TEXT PRIMARY KEY, val BLOB NOT NULL, " ~
+            "enrolled_height INTEGER, preimage BLOB)");
+    }
+
+    /***************************************************************************
+
+        Shut down the database
+
+        Note: this method must be called explicitly, and not inside of
+        a destructor.
+
+    ***************************************************************************/
+
+    public void shutdown ()
+    {
+        this.db.close();
+    }
+
+    /***************************************************************************
+
+        Add a enrollment data to the validators set
+
+        Params:
+            block_height = the current block height in the ledger
+            finder = the delegate to find UTXOs with
+            enroll = the enrollment data to add
+
+        Returns:
+            true if the enrollment data has been added to the validator set
+
+    ***************************************************************************/
+
+    public bool add (ulong block_height, scope UTXOFinder finder,
+        const ref Enrollment enroll) @safe nothrow
+    {
+        static ubyte[] buffer;
+
+        // check validity of the enrollment data
+        if (auto reason = isInvalidEnrollmentReason(enroll, block_height + 1,
+            finder))
+        {
+            log.info("Invalid enrollment data, Reason: " ~ reason, enroll);
+            return false;
+        }
+
+        // check if already exists
+        try
+        {
+            if (this.hasEnrollment(enroll.utxo_key))
+            {
+                log.info("Rejected already existing enrollment", enroll);
+                return false;
+            }
+        }
+        catch (Exception ex)
+        {
+            log.error("Exception occured in checking if the enrollment data exists", enroll, ex);
+            return false;
+        }
+
+        buffer.length = 0;
+        () @trusted { assumeSafeAppend(buffer); } ();
+
+        scope SerializeDg dg = (scope const(ubyte[]) data) nothrow @safe
+        {
+            buffer ~= data;
+        };
+
+        try
+        {
+            serializePart(enroll, dg);
+        }
+        catch (Exception ex)
+        {
+            log.error("Serialization error", enroll, ex);
```

Comment:
```suggestion
            log.error("Serialization error: {}, data was: {}", ex.msg, enroll);
```

As a rule of thumb, always put the message first. Because you always consider the data *in context of* the message.